File: web/src/stringify.ts
interface Status {
	indent: number;
	result: string;
	ansi: boolean;
	indentText: string;
}

interface Options {
	ansi?: boolean;
	indent?: string;
}

/**
 * Converts an object into a string containing Structured Data Notation.
 */
export default function stringify(obj: Record<PropertyKey, any>, options?: Options): string {
	let status = {
		indent: 0,
		result: "",
		ansi: options?.ansi ?? false,
		indentText: options?.indent ?? "\t",
	};

	printValue(obj, status);

	return status.result;
}

function printValue(obj: any, status: Status) {
	if (obj === null) {
		status.result += "null";
	} else if (Array.isArray(obj)) {
		status.result += "[";
		status.indent += 1;

		for (let i = 0; i < obj.length; i++) {
			status.result += "\n";
			indent(status);
			printValue(obj[i], status);
			if (i < obj.length - 1) {
				status.result += ",";
			}
		}

		status.indent -= 1;
		status.result += "\n";
		indent(status);
		status.result += "]";
	} else if (obj instanceof Date) {
		const dateStr = formatDate(obj);
		status.result += status.ansi ? `\x1b[35m${dateStr}\x1b[0m` : dateStr;
	} else if (typeof obj === "object") {
		status.result += "{";
		status.indent += 1;

		const keys = Object.keys(obj);
		for (let i = 0; i < keys.length; i++) {
			status.result += "\n";
			const key = keys[i];
			indent(status);
			status.result += key + ": ";
			printValue(obj[key], status);
			if (i < keys.length - 1) {
				status.result += ",";
			}
		}

		status.indent -= 1;
		status.result += "\n";
		indent(status);
		status.result += "}";
	} else if (typeof obj === "string") {
		status.result += status.ansi ? `\x1b[32m"${obj}"\x1b[0m` : `"${obj}"`;
	} else if (typeof obj === "number") {
		status.result += status.ansi ? `\x1b[33m${String(obj)}\x1b[0m` : String(obj);
	} else if (typeof obj === "boolean") {
		status.result += status.ansi ? `\x1b[34m${String(obj)}\x1b[0m` : String(obj);
	} else {
		status.result += String(obj);
	}
}

function indent(status: Status) {
	status.result += status.indentText.repeat(status.indent);
}

function formatDate(date: Date): string {
	const year = date.getFullYear();
	const month = String(date.getMonth() + 1).padStart(2, "0");
	const day = String(date.getDate()).padStart(2, "0");
	const hours = String(date.getHours()).padStart(2, "0");
	const minutes = String(date.getMinutes()).padStart(2, "0");
	const seconds = String(date.getSeconds()).padStart(2, "0");

	if (hours === "00" && minutes === "00" && seconds === "00") {
		return `${year}-${month}-${day}`;
	} else {
		return `${year}-${month}-${day}T${hours}:${minutes}`;
	}
}

File: zig/src/stringify.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const Value = @import("./types/Value.zig").Value;
const parse = @import("parse.zig");

/// Options for stringification
pub const Options = struct {
    ansi: bool = false,
    indent: []const u8 = "\t",
};

/// Status tracks stringification state
const Status = struct {
    indent: usize,
    result: std.ArrayListUnmanaged(u8),
    ansi: bool,
    indentText: []const u8,
    allocator: Allocator,

    pub fn init(allocator: Allocator, ansi: bool, indentText: []const u8) Status {
        return Status{
            .indent = 0,
            .result = .empty,
            .ansi = ansi,
            .indentText = indentText,
            .allocator = allocator,
        };
    }

    pub fn deinit(self: *Status) void {
        self.result.deinit(self.allocator);
    }
};

/// Converts a Value into a string containing Structured Data Notation
pub fn stringify(allocator: Allocator, value: Value, options: ?*const Options) ![]const u8 {
    const opts = options orelse &Options{};
    var status = Status.init(allocator, opts.ansi, opts.indent);
    defer status.deinit();

    try printValue(allocator, value, &status);

    return status.result.toOwnedSlice(allocator);
}

fn printValue(allocator: Allocator, value: Value, status: *Status) !void {
    switch (value) {
        .array => |arr| {
            try status.result.append(allocator, '[');
            status.indent += 1;

            for (arr.items, 0..) |item, i| {
                try status.result.append(allocator, '\n');
                indent(status);
                try printValue(allocator, item, status);
                if (i < arr.items.len - 1) {
                    try status.result.append(allocator, ',');
                }
            }

            status.indent -= 1;
            try status.result.append(allocator, '\n');
            indent(status);
            try status.result.append(allocator, ']');
        },
        .date => |d| {
            if (status.ansi) {
                const formatted = try std.fmt.allocPrint(allocator, "\x1b[35m{s}\x1b[0m", .{d});
                defer allocator.free(formatted);
                try status.result.appendSlice(allocator, formatted);
            } else {
                try status.result.appendSlice(allocator, d);
            }
        },
        .object => |obj| {
            try status.result.append(allocator, '{');
            status.indent += 1;

            var iter = obj.iterator();
            var keys: std.ArrayListUnmanaged([]const u8) = .empty;
            defer {
                for (keys.items) |k| {
                    allocator.free(k);
                }
                keys.deinit(allocator);
            }

            while (iter.next()) |entry| {
                const key_copy = try allocator.dupe(u8, entry.key_ptr.*);
                try keys.append(allocator, key_copy);
            }

            for (keys.items, 0..) |key, i| {
                try status.result.append(allocator, '\n');
                indent(status);
                const key_part = try std.fmt.allocPrint(allocator, "{s}: ", .{key});
                defer allocator.free(key_part);
                try status.result.appendSlice(allocator, key_part);
                try printValue(allocator, obj.get(key).?, status);
                if (i < keys.items.len - 1) {
                    try status.result.append(allocator, ',');
                }
            }

            status.indent -= 1;
            try status.result.append(allocator, '\n');
            indent(status);
            try status.result.append(allocator, '}');
        },
        .string => |s| {
            if (status.ansi) {
                const formatted = try std.fmt.allocPrint(allocator, "\x1b[32m\"{s}\"\x1b[0m", .{s});
                defer allocator.free(formatted);
                try status.result.appendSlice(allocator, formatted);
            } else {
                const formatted = try std.fmt.allocPrint(allocator, "\"{s}\"", .{s});
                defer allocator.free(formatted);
                try status.result.appendSlice(allocator, formatted);
            }
        },
        .int => |n| {
            if (status.ansi) {
                const formatted = try std.fmt.allocPrint(allocator, "\x1b[33m{d}\x1b[0m", .{n});
                defer allocator.free(formatted);
                try status.result.appendSlice(allocator, formatted);
            } else {
                const formatted = try std.fmt.allocPrint(allocator, "{d}", .{n});
                defer allocator.free(formatted);
                try status.result.appendSlice(allocator, formatted);
            }
        },
        .num => |n| {
            if (status.ansi) {
                const formatted = try std.fmt.allocPrint(allocator, "\x1b[33m{d}\x1b[0m", .{n});
                defer allocator.free(formatted);
                try status.result.appendSlice(allocator, formatted);
            } else {
                const formatted = try std.fmt.allocPrint(allocator, "{d}", .{n});
                defer allocator.free(formatted);
                try status.result.appendSlice(allocator, formatted);
            }
        },
        .bool => |b| {
            if (status.ansi) {
                const formatted = try std.fmt.allocPrint(allocator, "\x1b[34m{any}\x1b[0m", .{b});
                defer allocator.free(formatted);
                try status.result.appendSlice(allocator, formatted);
            } else {
                const formatted = try std.fmt.allocPrint(allocator, "{any}", .{b});
                defer allocator.free(formatted);
                try status.result.appendSlice(allocator, formatted);
            }
        },
        .null => {
            try status.result.appendSlice(allocator, "null");
        },
    }
}

fn indent(status: *Status) void {
    for (0..status.indent) |_| {
        status.result.appendSlice(status.allocator, status.indentText) catch unreachable;
    }
}
